Add original comment feature and error handle module

YuncongZhou opened the pull request:
User module development paused

YuncongZhou committed:
Bug fixing one

Add a parent_id validation block

YuncongZhou committed:
Bug fixing

Deal with an issue when parent id not found

YuncongZhou committed:
delete dependency

YuncongZhou committed:
delete files

congqiao commented on server/src/index.js:
```diff
@@ -33,24 +33,28 @@ const main = async () => {
-      const parentId = ObjectID.createFromHexString(req.body.parent_id, (err, r) => {
-        if (err) {
+      const re = /[0-9a-f]{24}/
+      if (!(req.body.parent_id.length === 24 && re.test(req.body.parent_id))) {
```
Use ```try...catch```.

congqiao commented on server/src/index.js:
```diff
@@ -33,24 +33,28 @@ const main = async () => {
+        console.log('Invalid parent id found. Abort.')
+        res.status(409).json({ success: false })
+      } else {
+        const parentId = ObjectID.createFromHexString(req.body.parent_id)
```
`req.body.parentId`

congqiao replied:
Allow comments with no parent ID.

congqiao commented on server/src/index.js:
```diff
@@ -33,24 +33,28 @@ const main = async () => {
+          await db.collection('Data').updateOne(
+            { _id: parentId },
+            { $push: { comment_ids: r.insertedId.toHexString() } },
+            (err, result) => {
```
Not necessary.

congqiao commented on server/src/index.js:
```diff
@@ -33,24 +33,28 @@ const main = async () => {
+          res.status(201).json({ success: true })
+        } else {
           console.log('Parent post not found. Abort.')
           res.status(409).json({ success: false })
```
`400`

congqiao replied:
`success` not necessary

YuncongZhou committed:
minor changes

Add a feature that allow the users publish a pure comment with no need
of url or parent post.

congqiao commented:
The title is cryptic and irrelevant.

congqiao commented on server/src/index.js:
```diff
@@ -27,30 +27,34 @@ const main = async () => {
         const r = await db.collection('Data').insertOne(data)
         console.log('News added to the database.')
-        res.status(201).json({ success: true })
+        res.send(201)
```
`sendStatus`. FFS read docs, dude!

congqiao commented on server/src/index.js:
```diff
@@ -27,30 +27,34 @@ const main = async () => {
-        res.status(201).json({ success: true })
+        res.send(201)
       } else {
         console.log('Duplicate news found. Abort.')
```
Same.

congqiao commented on server/src/index.js:
```diff
@@ -27,30 +27,34 @@ const main = async () => {
       if (!await db.collection('Data').findOne({ url: data.url })) {
         const r = await db.collection('Data').insertOne(data)
         console.log('News added to the database.')
```
Let's avoid `console.log`. You should rely on your debugger for debugging.

congqiao commented on server/src/index.js:
```diff
@@ -27,30 +27,34 @@ const main = async () => {
       } else {
         console.log('Duplicate news found. Abort.')
-        res.status(409).json({ success: false })
+        res.send(409)
```
You should return the existing ID so that the client can redirect the user to the page.

congqiao commented on server/src/index.js:
```diff
@@ -27,30 +27,34 @@ const main = async () => {
+          res.send(400)
+        }
+      } catch (err) {
+        if (req.body.parentId === null) {
```
This check (whether the comment has a parent) should be outside. What if `ObjectID.createFromHexString(null)` is not an error?

congqiao commented on server/src/index.js:
```diff
@@ -27,30 +27,34 @@ const main = async () => {
+      } catch (err) {
+        if (req.body.parentId === null) {
+          const r = await db.collection('Data').insertOne(data)
+          console.log('Comment added to the database.')
```
Same.

congqiao commented on server/src/index.js:
```diff
@@ -27,30 +27,34 @@ const main = async () => {
+        if (req.body.parentId === null) {
+          const r = await db.collection('Data').insertOne(data)
+          console.log('Comment added to the database.')
+          res.send(201)
```
Same.

congqiao commented on server/src/index.js:
```diff
@@ -27,30 +27,34 @@ const main = async () => {
+          console.log('Comment added to the database.')
+          res.send(201)
+        } else {
+          console.log('Invalid parent ID. Abort.')
```
Same.

congqiao commented on server/src/index.js:
```diff
@@ -27,30 +27,34 @@ const main = async () => {
+          res.send(201)
+        } else {
+          console.log('Invalid parent ID. Abort.')
+          res.send(400)
```
Same.

YuncongZhou committed:
Change return value and delete console log

Change the logic to determine an original comment
Return post id once published successfully
Send only status code once failed
Delete console logs

congqiao commented on server/src/index.js:
```diff
@@ -26,31 +26,31 @@ const main = async () => {
-        console.log('News added to the database.')
-        res.status(201).json({ success: true })
+        const id = r.insertedId.toHexString()
+        res.status(201).send({ post_id: id })
```
You use "data" and now "posts". You should be consistent. Let's avoid data since it is too generic and use posts.

congqiao commented on server/src/index.js:
```diff
@@ -26,31 +26,31 @@ const main = async () => {
     if (data.type === 'news') {
```
Use `switch` to figure out which type of post is out of the three types.

congqiao replied:
Also let's delay creating the new object which is to be inserted into the database until before insertion, since different types of posts have different fields.

congqiao commented on server/src/index.js:
```diff
@@ -26,31 +26,31 @@ const main = async () => {
       } else {
-        console.log('Duplicate news found. Abort.')
-        res.status(409).json({ success: false })
+        res.sendStatus(409)
```
Also send the existing post ID.

congqiao commented on server/src/index.js:
```diff
@@ -26,31 +26,31 @@ const main = async () => {
-      } else {
-        console.log('Parent post not found. Abort.')
-        res.status(409).json({ success: false })
+      } catch (err) {
```
Could you try if `catch ()` works?

YuncongZhou replied:
It works.

YuncongZhou committed:
Add switch block and function

Add a post creating function
Use switch to choose news and two kinds of comments
Change the fields in post

congqiao commented on server/src/index.js:
```diff
@@ -11,52 +11,83 @@ const url = process.env.URL
 app.listen(3000, () => console.log('Listening to port 3000'))
 app.get('/user/:id', (req, res) => res.send(`Welcome to the homepage of user ${req.params.id}.`))
 
+function createPost(req, type) {
```
Use the arrow function: `const createPost = (req, type) => ...`

congqiao replied:
This is a bad separation of concern. This function duplicates the logic of `switch (type)`. Also `type` is in `req.body`, which is a duplicate parameter. Third, `req` is an Express request object. This function should have no knowledge of Express.

Maybe put this back for now. We can think about reducing complexity later.

congqiao commented on server/src/index.js:
```diff
@@ -11,52 +11,83 @@ const url = process.env.URL
+      post.parent_id = req.body.parentId
+      break
+    default:
+      break
```
No need to `break` here.

congqiao commented on server/src/index.js:
```diff
@@ -11,52 +11,83 @@ const url = process.env.URL
+  const post = {}
+  switch (type) {
+    case 0:
+      post.title = req.body.title
```
You need to check if the title is valid, otherwise people can submit empty title. You can use `if (req.body.title)`. If it's invalid, you should return 400.

congqiao commented on server/src/index.js:
```diff
@@ -11,52 +11,83 @@ const url = process.env.URL
+  switch (type) {
+    case 0:
+      post.title = req.body.title
+      post.url = req.body.url
```
Same.

congqiao commented on server/src/index.js:
```diff
@@ -11,52 +11,83 @@ const url = process.env.URL
+      post.url = req.body.url
+      break
+    case 1:
+      post.title = req.body.title
```
Same.

congqiao commented on server/src/index.js:
```diff
@@ -11,52 +11,83 @@ const url = process.env.URL
+      break
+    case 1:
+      post.title = req.body.title
+      post.body = req.body.body
```
Same.

congqiao commented on server/src/index.js:
```diff
@@ -11,52 +11,83 @@ const url = process.env.URL
+      post.body = req.body.body
+      break
+    case 2:
+      post.body = req.body.body
```
Same.

congqiao commented on server/src/index.js:
```diff
@@ -11,52 +11,83 @@ const url = process.env.URL
-          console.log('Parent post not found. Abort.')
-          res.status(409).json({ success: false })
+      case 2: {
+        try {
```
You should make `try` block as small as possible. This catches a lot of exceptions, not necessarily from the statement you want. Here it's best to do
```javascript
let parent
try {
  parent = await db.collection('posts').findOne({ _id: parentId })
} ...
```

congqiao commented on server/src/index.js:
```diff
@@ -11,52 +11,83 @@ const url = process.env.URL
       }
+      default:
+        res.sendStatus(400)
+        break
```
No need to `break`.

YuncongZhou committed:
Change arrow function

Change arrow function
Delete break statement in default section

YuncongZhou committed:
Restructure switch logic

Create post body at the end of main program

congqiao commented on server/src/index.js:
```diff
@@ -11,52 +11,76 @@ const url = process.env.URL
-          res.status(409).json({ success: false })
+  // post case 0: news , case 1: top-level comment, case 2: non top-level comment
+  app.post('/posts', async (req, res) => {
+    let parentId = null
```
Do not assign a default value. `null` has a meaning.

YuncongZhou committed:
Delete parentId init value null
